backend_uml.go

```diff
@@ -0,0 +1,200 @@
+	b.moduleDir = base + "linux-uml/modules/lib/modules"
+	b.slirpHelperPath = base + "libslirp-rs/target/debug/libslirp-helper"
+	return b
+}
+
+func (b umlBackend) Name() string {
+	return "uml"
+}
+
+func (b umlBackend) Supported() (bool, error) {
+	// check the binaries are present
+	if _, err := os.Stat(b.kernelPath); err != nil {
+		return false, fmt.Errorf("user-mode-linux not installed")
+	}
+	if _, err := os.Stat(b.moduleDir); err != nil {
+		return false, fmt.Errorf("user-mode-linux not installed")
+	}
+	if _, err := os.Stat(b.slirpHelperPath); err != nil {
+		// TODO confirm package name
+		return false, fmt.Errorf("slirp-helper not installed")
+	}
+	return true, nil
+}
+
+func (b umlBackend) KernelPath() (string, string, error) {
+	// check the required paths exist
+	if _, err := b.Supported(); err != nil {
+		return "", "", err
+	}
+
+	return b.kernelPath, b.moduleDir, nil
+}
+
+func (b umlBackend) InitrdModules() []string {
+	return []string{}
+}
+
+func (b umlBackend) UdevRulesImageMatch() string {
+	return "ubd*"
+}
+
+func (b umlBackend) NetworkdMatch() string {
+	return "vec*"
+}
+
+func (b umlBackend) JobOutputTTY() string {
+	// Send the fakemachine job output to the right console
+	if b.machine.showBoot {
+		return "/dev/tty0"
+	}
+	return "/dev/tty1"
+}
+
+func (b umlBackend) MountParameters(mount mountPoint) (fstype string, options []string) {
+	fstype = "hostfs"
+	options = []string{mount.hostDirectory}
+	return
+}
+
+func (b umlBackend) InitModules() []string {
+	return []string{}
+}
+
+func (b umlBackend) InitExtraVolumes() []mountPoint {
+	// mount the UML modules over the top of /lib/modules
+	// which contains the modules from the base system
+	moduleVolume := mountPoint{hostDirectory: b.moduleDir, machineDirectory: "/lib/modules", label: "modules", static: true}
+	return []mountPoint{moduleVolume}
+}
+
+func (b umlBackend) Start() (int, error) {
+	// TODO commandWrapper forces network-wait-online & doesnt exit after network setup fail ?
+	// TODO disable lvm2-activation-generator ?
+	// TODO linux process is leftover when manually killing fakemachine. this bug also happens on KVM so should be tackled in follow-up
+	// TODO killing helper doesn't stop fakemachine
+
+	m := b.machine
+
+	// create a socketpair
+	// TODO syscall has been depreciated in favour of https://godoc.org/golang.org/x/sys
+	fds, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_DGRAM, 0)
+	if err != nil {
+		return -1, err
+	}
+
+	// f1 is attached to the slirp-helper
+	f1 := os.NewFile(uintptr(fds[0]), "")
+	if f1 == nil {
+		return -1, fmt.Errorf("socketpair: f1 is incorrect")
+	}
+	defer f1.Close()
+
+	// f2 is attached to the uml guest
+	f2 := os.NewFile(uintptr(fds[1]), "")
+	if f2 == nil {
+		return -1, fmt.Errorf("socketpair: f2 is incorrect")
+	}
+	defer f2.Close()
+
+
+	// launch the slirp helper and attach to socketpair using --fd=0 argument
+	slirp_args := []string{"libslirp-helper",
+				"--fd=3",
```

Comment:
is fd=3 something to actaully hardcode or should that be dynamic ?

Reply:
i am going to go for hardcode; i can't see it changing unless the helper args change

Reply:
But this is the fd you get back from the socketpair right? Which is not something you can assume will be 3

Reply:
the fd created by the socketpair is indeed variable, which is then attached to the process using `os.ProcAttr`.
The `--fd` argument tells the helper which fd which has been attached to the process to use. It's always attached after stdout, stderr etc. Hopefully that's a bit clearer ?

Reply:
That deserved a comment at the very least and probably moving this code below the ProcAttr setup as it's very unclear where this =3 comes from and it's ok to hardcode it like that

Reply:
right i've split the code up a bit and hopefully explained it better in the comments...